File: Problems/Algorithms/1162. As Far from Land as Possible/FarLand.kt
class Solution {
    fun maxDistance(grid: Array<IntArray>): Int {
        val n = grid.size
        val m = grid[0].size
        
        val neighbors = arrayOf(intArrayOf(0, 1), intArrayOf(0, -1), intArrayOf(1, 0), intArrayOf(-1, 0))
        val visited = Array(n) { BooleanArray(m) { false } }
        
        val queue = ArrayDeque<Pair<Int, Int>>()
        for (r in 0..n-1) {
            for (c in 0..m-1) {
                if (grid[r][c] == 1) {
                    queue.add(Pair(r, c))
                }
            }
        }
        
        var ans = -1
        while (!queue.isEmpty()){
            for (i in 1..queue.size) {
                val curr = queue.removeFirst()
                val x = curr.first
                val y = curr.second
                
                for (neighbor in neighbors) {
                    val r = x + neighbor[0]
                    val c = y + neighbor[1]
                    
                    if (isValid(n, m, r, c) && !visited[r][c] && grid[r][c] == 0) {
                        visited[r][c] = true
                        queue.add(Pair(r, c))
                    }
                }
            }
            ans++
        }
        
        if (ans <= 0) return -1
        return ans
    }
    
    private fun isValid(n: Int, m: Int, r: Int, c: Int): Boolean {
        return r >= 0 && r < n && c >= 0 && c < m
    }
}

File: Problems/Algorithms/1162. As Far from Land as Possible/far_land.py
class Solution:
    def maxDistance(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
    
        visited = [[False for _ in range(m)] for _ in range(n)]
        neighbors = [[0, 1], [0, -1], [1, 0], [-1, 0]]
    
        queue = []
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    queue.append((i, j))
    
        ans = -1
        while queue:
            size = len(queue)
            for _ in range(size):
                curr = queue.pop(0)
                
                for neighbor in neighbors:
                    r, c = curr[0] + neighbor[0], curr[1] + neighbor[1]
                    if r >= 0 and r < n and c >= 0 and c < m and not visited[r][c] and grid[r][c] == 0:
                        visited[r][c] = True
                        queue.append((r, c))

            ans += 1
    
        return -1 if ans <= 0 else ans
